src/Orleans.Core/Diagnostics/ActivityPropagationGrainCallFilter.cs

```diff
@@ -0,0 +1,277 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Orleans.Runtime
+{
+    internal static class ActivityPropagationGrainCallFilter
+    {
+        private const string TraceParentHeaderName = "traceparent";
+        private const string TraceStateHeaderName = "tracestate";
+        private const string TraceBaggageHeaderName = "tracebaggage";
+
+        internal const string ActivityNameIn = "Orleans.Runtime.GrainCall.In";
+        internal const string ActivityStartNameIn = "Orleans.Runtime.GrainCall.In.Start";
+        internal const string ExceptionEventNameIn = "Orleans.Runtime.GrainCall.In.Exception";
+
+        internal const string ActivityNameOut = "Orleans.Runtime.GrainCall.Out";
+        internal const string ActivityStartNameOut = "Orleans.Runtime.GrainCall.Out.Start";
+        internal const string ExceptionEventNameOut = "Orleans.Runtime.GrainCall.Out.Exception";
+
+        public class ActivityPropagationOutgoingGrainCallFilter : IOutgoingGrainCallFilter
+        {
+            private readonly DiagnosticListener diagnosticListener;
+
+            public ActivityPropagationOutgoingGrainCallFilter(DiagnosticListener diagnosticListener)
+            {
+                this.diagnosticListener = diagnosticListener;
+            }
+
+            public Task Invoke(IOutgoingGrainCallContext context)
+            {
+                if (Activity.DefaultIdFormat != ActivityIdFormat.W3C)
```

Comment:
Maybe it is better to move this check into `AddActivityPropagation` and throw an exception.

Reply:
Agreed. The exception message can inform the user that they need to configure this property beforehand